common/membership/hashring.go

```diff
@@ -265,7 +273,8 @@ func (r *ring) refreshRingWorker() {
 		select {
 		case <-r.shutdownCh:
 			return
-		case <-r.refreshChan: // local signal or signal from provider
+		case event := <-r.refreshChan: // local signal or signal from provider
+			r.notifySubscribers(event)
```

Comment:
why notify about "a refresh has been requested"?  doesn't seem related to what subscribers are observing

Reply:
hmm, this case is for notification from peer provider (ringpop by default). When ringpop sends a an event to refreshChan, we notify "service-level" subscribers.

Reply:
that's... rather trivially incorrect about what's happening in this code

Reply:
yea... the whole setup of writing to the notification channel seems to assume asynchronous work (non-blocking writes in a tight loop), so any notified subscriber would get a blend of old and new data if they call `Lookup(...)`.  that seems very unlikely to be desired.

e.g. the acquireShards subscriber does exactly that, and calls `Lookup(..)` for all shards, racing with the actual new-data update.

Reply:
Good point about the order of operations here. It looks like if we do `r.notifySubscribers(event)` after `r.refresh()` then the acquireShards will do `Lookup()` on new data.

I also just checked the shard controller's [membershipUpdateCh](https://github.com/uber/cadence/blob/5003ffbff3d7ee7ba430b8bc601865448650c265/service/history/shard/controller.go#L128) which is initialized with 10 capacity. So notifications will most likely be unblocked but acquiring shards for each change might cause too many redundant work. Since r.refresh() already has [4s cooldown](https://github.com/uber/cadence/blob/ff95842fb10d2d9630bbd1665d3d02430b447d9c/common/membership/hashring.go#L231), it should return a bool to indicate whether a refresh happened and then only notify subscribers if refresh occurred

Reply:
now we check if ring has actually changed and only after that subscribers are notified